### PR TITLE
Reconfigure Consensus

### PR DESCRIPTION
**Pull request by asonnino**
Allow the AuthorityState to make decide when to reconfigure consensus. In a nutshell, this is how we plan to reconfigure consensus in Sui:

* We have a watcher that lists for consensus updates, gracefully shutdowns all tasks, and then gracefully reboots them: https://github.com/MystenLabs/narwhal/blob/32b5298b60d410620f2c0e80eec4b6c12e4f9bdd/node/src/restarter.rs#L20
* Here is an end-to-end test where SimpleExecutionError is supposed to represent AuthorityState in Sui: https://github.com/MystenLabs/narwhal/blob/32b5298b60d410620f2c0e80eec4b6c12e4f9bdd/node/tests/reconfigure.rs#L162

Note that this strategy will not work for our external partners (Celo, Somelier) so the Narwhal codebase has other ways to change epoch/reconfigure.

### CODE REVIEW

File: crates/sui-core/src/authority_server.rs [hunk 1]
```diff
@@ -178,27 +181,50 @@ impl ValidatorService {
...
+        // )
+        // .await?;
+        // narwhal_node::Node::spawn_workers(
+        //     consensus_name,
+        //     /* ids */ vec![0], // We run a single worker with id '0'.
+        //     config.genesis()?.narwhal_committee(),
+        //     &consensus_store,
+        //     consensus_config.narwhal_config().to_owned(),
+        //     prometheus_registry,
+        // );
```
Comment (mystenmark) [inline crates/sui-core/src/authority_server.rs:227]: commented code?

Comment (asonnino) [reply crates/sui-core/src/authority_server.rs:227]: oops

Comment (asonnino) [reply crates/sui-core/src/authority_server.rs]: fixed

**Comment by lxfind** [comment]
Cool! Could you try to see if you are able to connect this to here https://github.com/MystenLabs/sui/blob/4c5877da019b4f074046b0d9d0df01d6d246461a/crates/sui-core/src/epoch/reconfiguration.rs#L162 ?

**Review by lxfind — APPROVED** [review]

**Comment by asonnino** [comment]
> Cool! Could you try to see if you are able to connect this to here
> 
> https://github.com/MystenLabs/sui/blob/4c5877da019b4f074046b0d9d0df01d6d246461a/crates/sui-core/src/epoch/reconfiguration.rs#L162
> 
> ?

https://github.com/MystenLabs/sui/issues/4302

**Pull request merged**